File: LeetCodeExample.Test/DP/_00198_HouseRobber.py
# You are a professional robber planning to rob houses along a street. Each house has a certain amount of money stashed, the only constraint stopping you from robbing each of them is that adjacent houses have security systems connected and it will automatically contact the police if two adjacent houses were broken into on the same night.

# Given an integer array nums representing the amount of money of each house, return the maximum amount of money you can rob tonight without alerting the police.

class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return max(nums[0], nums[1])
        dp = [0] * len(nums)
        #dp = [[0 for i in range(2)] for i in range(len(nums))]
        #dp[0][0] = nums[0]
        #dp[1][0] = nums[1]
        #dp[1][1] = nums[0]
        dp[0] = nums[0]
        dp[1] = max(nums[0], nums[1])

        for i in range(2, len(nums)):
            dp[i] = max(dp[i - 2] + nums[i], dp[i - 1])
            #dp[i][0] = max(dp[i - 2][0], dp[i - 2][1]) + nums[i]
            #dp[i][1] = max(dp[i - 1][0], dp[i - 1][1])
        return dp[len(nums) - 1]

File: LeetCodeExample.Test/DP/_00198_HouseRobber.cs
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Linq;

namespace LeetCodeExample.Test
{
    /// <summary>
    // You are a professional robber planning to rob houses along a street.Each house has a certain amount of money stashed, the only constraint stopping you from robbing each of them is that adjacent houses have security systems connected and it will automatically contact the police if two adjacent houses were broken into on the same night.
    // Given an integer array nums representing the amount of money of each house, return the maximum amount of money you can rob tonight without alerting the police.
    /// </summary>
    public class _00198_HouseRobber
    {
        public int Rob(int[] nums)
        {
            // One extra space used to remove edge cases and if statements
            int[] dp = new int[nums.Length + 1];
            dp[0] = 0;
            dp[1] = nums[0];
            for (int i = 2; i < nums.Length + 1; i++)
            {
                dp[i] = Math.Max(dp[i - 1], dp[i - 2] + nums[i - 1]);
            }

            return dp[nums.Length];
        }

        /*Dictionary<int, int> map = new Dictionary<int, int>();

        public int dp(int[] nums, int i)
        {
            //Console.WriteLine($"{i}");
            // Base case
            if (i == nums.Length - 1)
                return nums[i];
            if (i > nums.Length - 1)
                return 0;

            if (map.ContainsKey(i))
                return map[i];

            int maxProfit = 0;
            // Must skip the adjacent house
            for (int j = i + 2; j < nums.Length; j++)
            {
                int profit = dp(nums, j);
                maxProfit = Math.Max(maxProfit, profit);
            }
            var thisProfit = maxProfit + nums[i];

            map.Add(i, thisProfit);
            //Console.WriteLine($"{i} {thisProfit}");
            return thisProfit;
        }*/
    }
}